### ISSUE TITLE
<istream> : Why only readsome() need to set failbit when sentry is not true?

### ISSUE THREAD

**Issue by Kamichanw** [microsoft/STL#1592]
Here is implementation of standard library:
```
streamsize __CLR_OR_THIS_CALL readsome(_Elem* _Str, streamsize _Count) { 
        ios_base::iostate _State = ios_base::goodbit;
        _Chcount                 = 0;
        const sentry _Ok(*this, true);
        streamsize _Num;
        if (!_Ok) {
            _State |= ios_base::failbit; // no buffer, fail
        } else if ((_Num = _Myios::rdbuf()->in_avail()) < 0) {
            _State |= ios_base::eofbit; // no characters available
        } else if (0 < _Count && 0 < _Num) { // read available
            read(_Str, _Num < _Count ? _Num : _Count);
        }
        _Myios::setstate(_State);
        return gcount();
    }
```
In constructor of sentry, if `this->good() == false`, failbit will be set directly, so why `readsome()`  still need to set failbit ? and why only `readsome()` need to do so?

**Comment by MattStephanson** [comment]
In general, the behavior of an _unformatted input function_ is "if [...] the sentry object returns false, [...] the function returns without attempting to obtain any input" [[istream.unformatted]/1](http://eel.is/c++draft/input.output#istream.unformatted-1). `readsome` has additional requirements in [[istream.unformated]/31](http://eel.is/c++draft/input.output#istream.unformatted-31): "after constructing a sentry object, if !good() calls setstate(failbit) which may throw an exception, and return". Other functions also have this additional behavior, for example `read` and `putback`.

I could be missing something, but this looks redundant to me too. But the Standard explicitly describes this behavior, so it's usually best to follow it.

**Comment by Kami-chanw** [comment]
> In general, the behavior of an _unformatted input function_ is "if [...] the sentry object returns false, [...] the function returns without attempting to obtain any input" [[istream.unformatted]/1](http://eel.is/c++draft/input.output#istream.unformatted-1). `readsome` has additional requirements in [[istream.unformated]/31](http://eel.is/c++draft/input.output#istream.unformatted-31): "after constructing a sentry object, if !good() calls setstate(failbit) which may throw an exception, and return". Other functions also have this additional behavior, for example `read` and `putback`.
> 
> I could be missing something, but this looks redundant to me too. But the Standard explicitly describes this behavior, so it's usually best to follow it.

Thank you for your answer. So,  is the purpose of calling `setstate(failbit)` to try to throw an exception(if `exceptions()` be set before)?

**Comment by MattStephanson** [comment]
> Thank you for your answer. So, is the purpose of calling `setstate(failbit)` to try to throw an exception(if `exceptions()` be set before)?

I don't think so, because the `sentry` will have already called `setstate(failbit)` on the stream ([[istream.sentry]/2](http://eel.is/c++draft/input.output#istream.sentry-2)).

**Comment by Kami-chanw** [comment]
You are right, but I still cannot understand why Standard requires `readsome()` should call `setstate(failbit)` if `!good()`. After all, sentry will have already called `setstate(failbit)`

**Comment by Kami-chanw** [comment]
> > Thank you for your answer. So, is the purpose of calling `setstate(failbit)` to try to throw an exception(if `exceptions()` be set before)?
> 
> I don't think so, because the `sentry` will have already called `setstate(failbit)` on the stream ([[istream.sentry]/2](http://eel.is/c++draft/input.output#istream.sentry-2)).

You are right, but I still cannot understand why Standard requires readsome() should call setstate(failbit) if !good(). After all, sentry will have already called setstate(failbit)

**Comment by AraHaan** [comment]
I agree it seems a little too redundant and possibly could be a waste of a few clock cycles.

**Comment by MattStephanson** [comment]
The language in [istream.unformated]/31 et al. came first. You can see that in LWG-60 based on all the resolutions to 'Change the first sentence of the Effects clause from "If !good() calls" to "Behaves as an unformatted input function..."'.

Later, in LWG-419, there was discussion of whether the `sentry` or the extractor should set `failbit` if `eofbit` is already set. The conclusion was that the `sentry` should do it, which is when "calls `is.setstate(failbit)`" was added to [istream.sentry], not just for EOF but for `!good()` in general. I would guess the Committee just overlooked that it made the extractor behavior redundant.
